cpp/src/command_classes/WakeUp.cpp

```diff
@@ -457,23 +500,28 @@ void WakeUp::SendPending
 	Node* node = GetNodeUnsafe();
 	if( node != NULL )
 	{
-
 		if( !node->AllQueriesCompleted() )
 		{
 			sendToSleep = false;
 		}
 	}
 
 	/* if we are reloading, the QueryStage_Complete will take care of sending the device back to sleep */
-	if( sendToSleep && !reloading)
+	if( sendToSleep && !reloading )
 	{
-		Msg* msg = new Msg( "WakeUpCmd_NoMoreInformation", GetNodeId(), REQUEST, FUNC_ID_ZW_SEND_DATA, true );
-		msg->Append( GetNodeId() );
-		msg->Append( 2 );
-		msg->Append( GetCommandClassId() );
-		msg->Append( WakeUpCmd_NoMoreInformation );
-		msg->Append( GetDriver()->GetTransmitOptions() );
-		GetDriver()->SendMsg( msg, Driver::MsgQueue_WakeUp );
+		if( !allowDelay || m_delayNoMoreInfo == 0 ) {
+			Msg* msg = new Msg( "WakeUpCmd_NoMoreInformation", GetNodeId(), REQUEST, FUNC_ID_ZW_SEND_DATA, true );
+			msg->Append( GetNodeId() );
+			msg->Append( 2 );
+			msg->Append( GetCommandClassId() );
+			msg->Append( WakeUpCmd_NoMoreInformation );
+			msg->Append( GetDriver()->GetTransmitOptions() );
+			GetDriver()->SendMsg( msg, Driver::MsgQueue_WakeUp );
+		} else {
+			Log::Write( LogLevel_Info, GetNodeId(), "  Node %d has delayed sleep of %dms", GetNodeId(), m_delayNoMoreInfo );
+			TimerThread::TimerCallback callback = TT_STDBIND(&WakeUp::SendPending, this, false);
+			GetDriver()->GetTimer()->TimerSetEvent(m_delayNoMoreInfo, callback);
+		}
```

Comment:
hrm. For clarity, I'd split this out into a new method that just sends the NoMoreInfo:
Something like:
WakeUp::SendNoMoreInfo() {
Msg *msg = new Msg(...)
...
GetDriver()->SendMsg(...)
}

then here just do a:
if ( !allowDelay || m_delayNoMoreInfo ==0 ) {
SendNoMoreInfo()
} else {
TimerThread::TimerCallback callback = TT_STDBIND(&WakeUp::SendNoMoreInfo...);
...
}

Otherwise I'm scared if a application keeps queuing up messages, we could end up in a infinite loop in SendPending